imap/autocreate.c

```diff
@@ -93,330 +93,399 @@ static const char *get_script_name(const char *filename)
+            syslog(LOG_ERR, "autosieve_create: Error parsing script %s:%m.",
+                   source_script);
+            if (err && *err) {
+                syslog(LOG_ERR, "autosieve_create: %s.", err);
+                free(err);
+            }
+            goto failed2;
+        } /* if (sieve_script_parse_only()) */
 
-            unlink(sieve_bctmpname);
+        /* Generate Bytecode */
+        if (sieve_generate_bytecode(&bc, s) == TIMSIEVE_FAIL) {
+            syslog(LOG_ERR, "autocreate_sieve: problem compiling sieve script.");
             fclose(in_stream);
-            close(out_fd);
-            return 1;
-        }
+            goto failed2;
+        } /* if (sieve_generate_bytecode()) */
 
-        /* generate the bytecode */
-        if(sieve_generate_bytecode(&bc, s) == TIMSIEVE_FAIL) {
-            syslog(LOG_WARNING,"autocreate_sieve: problem compiling sieve script");
-            /* removing the copied script and cleaning up memory */
-            unlink(sieve_bctmpname);
-            sieve_script_free(&s);
+        if (sieve_emit_bytecode(out_fd, bc) == TIMSIEVE_FAIL) {
+            syslog(LOG_ERR, "autocreate_sieve: problem emitting sieve script.");
             fclose(in_stream);
-            close(out_fd);
-            return 1;
-        }
-
-        if(sieve_emit_bytecode(out_fd, bc) == TIMSIEVE_FAIL) {
-            syslog(LOG_WARNING,"autocreate_sieve: problem emitting sieve script");
-            /* removing the copied script and cleaning up memory */
-            unlink(sieve_bctmpname);
             sieve_free_bytecode(&bc);
             sieve_script_free(&s);
-            fclose(in_stream);
-            close(out_fd);
-            return 1;
-        }
+            goto failed2;
+        } /* if (sieve_emit_bytecode()) */
 
-        /* clean up the memory */
         sieve_free_bytecode(&bc);
         sieve_script_free(&s);
-    }
+    } /* if (do_compile) */
 
     xclose(out_fd);
     rewind(in_stream);
 
-    /* Copy the initial script */
+    /* Copy the source script */
     oldmask = umask(077);
-    if((out_fp = fopen(sieve_tmpname, "w")) == NULL) {
-        syslog(LOG_WARNING,"autocreate_sieve: Unable to open destination sieve script %s: %m", sieve_tmpname);
-        unlink(sieve_bctmpname);
-        umask(oldmask);
+
+    if ((out_fp = fopen(script_names.tmpname1, "w")) == NULL) {
+        syslog(LOG_ERR, "autocreate_sieve: Unable to open destination sieve"
+               "script %s: %m", script_names.tmpname1);
         fclose(in_stream);
-        return 1;
+        umask(oldmask);
+        goto failed2;
     }
     umask(oldmask);
 
-    while((r = fread(buf,sizeof(char), sizeof(buf), in_stream)) > 0) {
-        if( fwrite(buf,sizeof(char), r, out_fp) != (unsigned)r) {
-            syslog(LOG_WARNING,"autocreate_sieve: Problem writing to sieve script file %s: %m",sieve_tmpname);
+    while ((r = fread(buf, sizeof(char), sizeof(buf), in_stream)) > 0) {
```

Comment:
There indeed seems to be a lot of stuff in `lib/util.c`, sorry for not noticing it earlier. I shall use a few of them (here and in other places as well.)